### PR TITLE
18537 Amalgamation application schema

### PR DESCRIPTION
**Pull request by chenhongjing**
*Issue #:* /bcgov/entity#18537

*Description of changes:*
- added amalgamation application schema
- added unit tests
- updated version = 2.18.18

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the business-schemas license (Apache 2.0).

### CODE REVIEW

File: src/registry_schemas/example_data/schema_data.py [hunk 1]
```diff
@@ -85,6 +85,165 @@
...
+    'orderDate': '2021-01-30T09:56:01+08:00',
+    'effectOfOrder': 'planOfArrangement',
+    'orderDetails': 'A note about order',
+    'fileKey': '011e332d-1b8e-4218-8710-ad8ac1fbc592.pdf'
+}
+
+FOREIGN_JURISDICTION = {
+    'country': 'CA',
+    'region': 'AB'
+}
```
Comment (severinbeauvais) [inline src/registry_schemas/example_data/schema_data.py:99]: Note that region is optional (only used if country is Canada).

Also note that region could be "Federal".

File: src/registry_schemas/example_data/schema_data.py [hunk 2]
```diff
@@ -85,6 +85,165 @@
...
+        {
+            'role': 'amalgamating',
+            'identifier': 'BC1234567'
+        },
+        {
+            'role': 'amalgamating',
+            'legalName': 'Foreign Co.',
+            'foreignJurisdiction': FOREIGN_JURISDICTION,
+            'corpNumber': '123456'
+        }
```
Comment (severinbeauvais) [inline src/registry_schemas/example_data/schema_data.py:113]: Yes, I think both objects here are correct.

For foreign companies, we store the 3 data fields that staff entered (+ role).

For BC companies, we store just their identifier (+ role) -- we can look up everything else.

File: src/registry_schemas/schemas/amalgamation.json [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        "amalgamation": {
+            "type":"object",
+            "required": [
+                "type",
+                "amalgamatingBusinesses",
+                "nameRequest",
+                "offices",
+                "parties",
+                "contactPoint"
+            ],
```
Comment (severinbeauvais) [inline src/registry_schemas/schemas/amalgamation.json:20]: `courtApproval` is also required.

File: src/registry_schemas/schemas/amalgamation.json [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+                "shareStructure": {
+                    "$ref": "https://bcrs.gov.bc.ca/.well_known/schemas/share_structure"
+                },
+                "courtOrder": {
+                    "$ref": "https://bcrs.gov.bc.ca/.well_known/schemas/court_order#/properties/courtOrder"
+                },
+                "courtApproval": {
+                    "title": "Amalgamation statement",
+                    "type": "boolean"
+                }
```
Comment (argush3) [inline src/registry_schemas/schemas/amalgamation.json:100]: courtApproval should allow null value too

Comment (argush3) [reply src/registry_schemas/schemas/amalgamation.json]: @severinbeauvais the FE wants to be able to store a null for drafts for `courtApproval` right?

Comment (severinbeauvais) [reply src/registry_schemas/schemas/amalgamation.json]: In a draft amalg filing, yes, this value could be null (or the property doesn't exist).

In the filed JSON, this property must be present and either be True or False (never null).

Comment (argush3) [reply src/registry_schemas/schemas/amalgamation.json]: ok, ignore what I said then @chenhongjing.  

We can remove the null option as null won't be there when we validate.

File: src/registry_schemas/example_data/schema_data.py (hunk 1 above)
Comment (severinbeauvais) [reply src/registry_schemas/example_data/schema_data.py:99]: ^^ this is already captured in the foreign jurisdiction schema :)

File: src/registry_schemas/schemas/amalgamation.json [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+                                    "role",
+                                    "identifier"
+                                ]
+                            },
+                            {
+                                "required": [
+                                    "role",
+                                    "foreignJurisdiction",
+                                    "legalName",
+                                    "corpNumber"
```
Comment (argush3) [inline src/registry_schemas/schemas/amalgamation.json:47]: do we want to prefix `legalName` and `corpNumber` with `foreign`?

File: src/registry_schemas/schemas/amalgamation.json [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+                                    "holding",
+                                    "amalgamating"
+                                ]
+                            },
+                            "identifier": { "type": "string" },
+                            "foreignJurisdiction": { 
+                                "$ref": "https://bcrs.gov.bc.ca/.well_known/schemas/foreign_jurisdiction" 
+                            },
+                            "legalName": { "type": "string" },
+                            "corpNumber": { "type": "string" }
```
Comment (argush3) [inline src/registry_schemas/schemas/amalgamation.json:64]: do we want to prefix `legalName` and `corpNumber` with `foreign`?

**Comment by argush3** [comment]
do we need to support name translations like IA filing?  

image

File: src/registry_schemas/schemas/amalgamation.json (hunk 6 above)
Comment (severinbeauvais) [reply src/registry_schemas/schemas/amalgamation.json:64]: The UI doesn't need to -- it uses a union type like this, so the properties are kept separate by the language.

image

**Comment by severinbeauvais** [comment]
> do we need to support name translations like IA filing?

Yes, good catch!

File: src/registry_schemas/schemas/amalgamation.json (hunk 6 above)
Comment (argush3) [reply src/registry_schemas/schemas/amalgamation.json:64]: I think it makes sense for the schema definition to prefix `legalName` and `corpNumber` with `foreign` for clarity though.

So it sounds like the FE will be fine if we change the naming of these two fields?

Comment (chenhongjing) [reply src/registry_schemas/schemas/amalgamation.json:64]: A question, is there any possibility/need FE will use `legalName` for non-foreign companies as well? If so, `legalName` will be shared like `role`.

Comment (severinbeauvais) [reply src/registry_schemas/schemas/amalgamation.json]: No, UI will not use `legalName` except for foreign companies.

The UI would need to be updated to prefix those properties with `foreign` but it's under development right now so not big deal, but I still think it's not necessary -- we should use the typing tools at hand to distinguish object types.

Comment (argush3) [reply src/registry_schemas/schemas/amalgamation.json]: ok, @chenhongjing forget about the `foreign` prefix.

Comment (severinbeauvais) [reply src/registry_schemas/schemas/amalgamation.json]: Hahaha...

**Review by severinbeauvais — APPROVED** [review]
Looks good. Thanks for all the work and thought you put into this! :)

**Comment by argush3** [comment]
You will need to update `filing.json` too.

https://github.com/bcgov/business-schemas/blob/31e9a18ecc8dcd4a0598e9123a77fdcad6076cb6/src/registry_schemas/schemas/filing.json

image

**Review by severinbeauvais — APPROVED** [review]

**Review by argush3 — APPROVED** [review]

File: src/registry_schemas/example_data/schema_data.py [hunk 7]
```diff
@@ -85,6 +85,168 @@
     'deliveryInstructions': 'some delivery instructions'
 }
 
+COURT_ORDER = {
```
Comment (thorwolpert) [inline src/registry_schemas/example_data/schema_data.py:88]: Any reason why this was moved?

File: src/registry_schemas/schemas/filing.json [hunk 8]
```diff
@@ -223,7 +223,11 @@
                 "properties": {
                   "name": {
                     "not": {
-                      "enum": ["registration", "incorporationApplication"]
+                      "enum": [
+                        "registration",
+                        "incorporationApplication",
+                        "amalgamationApplication"
```
Comment (thorwolpert) [inline src/registry_schemas/schemas/filing.json:229]: Do we not want the business header block for Amalgamations?

**Review by thorwolpert — COMMENTED** [review]
Just a couple of questions.

**Review by thorwolpert — APPROVED** [review]

**Pull request merged**